crates/datasources/src/sqlserver/mod.rs

```diff
@@ -0,0 +1,559 @@
+    config: tiberius::Config,
+}
+
+impl SqlServerAccess {
+    /// Create access configuration from an ADO formatted connection string.
+    ///
+    /// ADO connection strings: <https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/connection-strings>
+    /// Example: "server=tcp:localhost,1433;user=SA;password=<YourStrong@Passw0rd>;IntegratedSecurity=true;TrustServerCertificate=true"
+    pub fn try_new_from_ado_string(conn_str: &str) -> Result<Self> {
+        let config = tiberius::Config::from_ado_string(conn_str)?;
+        Ok(Self { config })
+    }
+
+    /// Validate that we can connect to server.
+    pub async fn validate_access(&self) -> Result<()> {
+        let _state = SqlServerAccessState::connect(self.config.clone()).await?;
+        Ok(())
+    }
+
+    /// Validate that we can connect to a specific table.
+    pub async fn validate_table_access(&self, schema: &str, table: &str) -> Result<()> {
+        let state = SqlServerAccessState::connect(self.config.clone()).await?;
+        let _schema = state.get_table_schema(schema, table).await?;
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct SqlServerAccessState {
+    client: Client,
+    /// Handle for underlying sql server connection.
+    ///
+    /// Kept on struct to avoid dropping the connection.
+    _conn_handle: JoinHandle<()>,
+}
+
+impl SqlServerAccessState {
+    async fn connect(config: tiberius::Config) -> Result<Self> {
+        let socket =
+            match tokio::time::timeout(CONNECTION_TIMEOUT, TcpStream::connect(config.get_addr()))
+                .await
+            {
+                Ok(result) => result,
+                Err(_) => {
+                    return Err(SqlServerError::String(format!(
+                        "timed out connection to SQL Server after {} seconds",
+                        CONNECTION_TIMEOUT.as_secs(),
+                    )))
+                }
+            }?;
+        socket.set_nodelay(true)?;
+        let (client, connection) = client::connect(config, socket.compat_write()).await?;
+
+        let handle = tokio::spawn(async move {
+            if let Err(e) = connection.run().await {
+                warn!(%e, "sql server connection errored");
+            }
+        });
+
+        Ok(SqlServerAccessState {
+            client,
+            _conn_handle: handle,
+        })
+    }
+
+    /// Get the arrow schema for a table.
+    async fn get_table_schema(&self, schema: &str, name: &str) -> Result<ArrowSchema> {
+        let mut query = self
+            .client
+            .query(format!("SELECT * FROM {schema}.{name} WHERE 1=0"))
+            .await?;
+        let cols = query.columns().await?;
+
+        let cols = match cols {
+            Some(cols) => cols,
+            None => {
+                return Err(SqlServerError::String(
+                    "unable to determine schema for table, query returned no columns".to_string(),
+                ))
+            }
+        };
```

Comment:
shouldn't we use the information_schema instead of directly querying the table for the schema?

Reply:
https://learn.microsoft.com/en-us/sql/relational-databases/system-information-schema-views/columns-transact-sql?view=sql-server-ver16

Reply:
We could, but something we've faced before with the postgres stuff is what happens if you don't have access to information_schema? Also `DATA_TYPE` is a string, and I'd prefer not to match on a string.

Reply:
good point, didn't think about the access aspect.